### ISSUE TITLE
Better support for non-empty string

### ISSUE THREAD

**Issue by ottokruse** [nvie/decoders#434]
First off, great library!

To guard against a non-empty string (in TypeScript) I have to do this now:

    regex(/./, 'Field is required')

Q: Is there a cleaner way to do this?

I don't really like having to use a regexp for this -  yes it is quite simple, but maybe there is (/can be invented) an even simpler/idiomatic way to do this. Empty strings are almost never useful in real-life JSON.

Note that the message is mandatory in the type definition of the regex guard, which is kinda annoying for my use, as I'm putting "Field is required" there (all my string fields use this regex guard, I never want empty strings in my JSON).

Maybe I'm missing something, interested to hear what you think.

**Comment by nvie** [comment]
Thanks!

If you want to avoid the regex decoder, you can use a predicate to put extra constraints on the value.  See also https://github.com/nvie/decoders#adding-predicates.

Here's an example:

```
import { compose, predicate, string } from 'decoders';

const nonEmptyString: Decoder<string> = compose(
  string,
  predicate(s => s.length > 0, 'Must not be empty'),
);
```

(But you could also extend that to include white-space only strings, if you'd use the predicate `s => s.trim().length > 0` for example.)

**Comment by nvie** [comment]
Or... define a helper if negative-regexes are a common pattern you need:

```js
function nonMatch(pattern: RegExp, message: string): Decoder<string> {
  return compose(
    string,
    predicate(s => !pattern.test(s), message)
  );
}
```

And then use that helper to define your non-empty test:

```js
const nonEmptyString: Decoder<string> = nonMatch(/^\s*$/, 'Must be not be empty');
const friendly: Decoder<string> = nonMatch(/^(swear|words|go|here)$/, 'Must not be a swear word');
```

**Comment by ottokruse** [comment]
Cool!

Nice to know now some other options.

Would it make sense to be able to do this:

`
import { nonEmptyString } from 'decoders';
`

Kinda feel it is such a regular use case, it would be good to put in the library.

**Comment by j-f1** [comment]
Maybe as `string.nonEmpty`?

**Comment by nvie** [comment]
Yeah, I'd love to add `nonEmptyString` to the standard decoder set.  Will do very soon! 👍

**Comment by nvie** [comment]
> Maybe as `string.nonEmpty`?

From a typing perspective don't prefer `string.nonEmpty` (although it looks nice).  Since all decoder instances are of the type `Decoder<T>` (which is basically a function type), it would require either all decoders to grow a `nonEmpty` method, or make the string decoder a special case.  In the interest of composability, I'd prefer to keep the decoder API as simple as possible.

**Comment by nvie** [comment]
Thanks for opening this issue! This should be fixed and available in [v1.18.0](https://github.com/nvie/decoders/releases/tag/v1.18.0) now.
